### ISSUE TITLE
small change request

### ISSUE THREAD

**Issue by Perecli** [i8beef/Ecobee#6]
I'm using your API is a progressive web app. It has a web component and an automation component on the back end. This means that for every set of ecobee auth tokens there are at least two clients operating concurrently on them from separate processes. The way the API works right now is that once you initiate the client it assumes that it has sole control of the auth tokens but in my case there are several parties that constantly refresh the tokens and must be aware of each others changes otherwise the ecobee servers will complain of expired auth token submission. For this reason I maintain the auth tokens in a database that is accessed by all the parties. But for this to work the Client class has to allow mid stream token updates not just on instantiation of the class through the constructor. Of course I could create a new Client class every request but that would not be proper considering all the setup that goes on inside.

**Comment by Perecli** [comment]
Please, see pull request for changes made that solve this issue and a few others.

**Comment by i8beef** [comment]
So everything in your PR (https://github.com/i8beef/Ecobee/pull/8) looks ok (besides the lack of a line break between the return and the code on line 104 ;-) ), the only question I really have is given the async nature of most of this library, does it make sense to make the `Func` and `Action` a `Func<Task<StoredAuthToken>>` and `Func<Task>` and add proper awaits everywhere. It's easier to expose something that's sync as an async than it is to expose something async as sync.

Thoughts?

**Comment by Perecli** [comment]
Are you saying that _getStoredAuthTokenFunc should be returning a Task of StoredAuthToken rather than the StoredAuthToken itself. And _setStoredAuthTokenFunc should be converted from Action to Func and return a Task? Basically make the two lambdas async?

Is so then, yes that is a good catch. I overlooked that since I happen to not be doing any async work in there but somebody very well might.

**Comment by i8beef** [comment]
Yes, that's what I tried to say but GitHub ate my formatting a bit.

Also, please follow the passing of cancellation token where applicable (not sure off the top of my head what that looks like with a passed delegate), and ConfigureAwait(false) on any awaited call.

**Comment by Perecli** [comment]
Made the changes.
Do you know why on line 216, I get an error if I add ".ConfigureAwait(false)"
I don't get it on the other Invoke calls.

**Comment by i8beef** [comment]
Not sure. Maybe try doing an explicit null check and avoiding the null propagation?

Thought being if its null its continuing on to null.ConfigureAwait(false).

You might be able to add another null propagation before the ConfigureAwait(False), but I'd probably just go ahead and wrap that with a null check at this point as that line is getting long.

**Comment by Perecli** [comment]
Yep, that worked and I updated. Any other problems you see with this PR?

**Comment by i8beef** [comment]
I don't think so. I've been researching on the CancellationToken thing to see how common it is to present that in passed delegate like this... I see mixed stuff, but I think I'll have you keep it for now, as I can't really think of any better way to hide the detail.

**Comment by Perecli** [comment]
I think its fine. I've seen it implemented like this before and I don't see a better way either.

**Comment by i8beef** [comment]
#8

**Comment by i8beef** [comment]
Merged, should be up momentarily on NuGet. Feel free to close this when you're good.

**Comment by Perecli** [comment]
Any thoughts about the code I suggested for implementing timeouts in the other closed issue? Do you think I am going down the right path or you have other ideas?

Ecobee's servers are down so often that I think this is needed.

**Comment by Perecli** [comment]
Actually I have something new in mind using "CancellationTokenSource.CreateLinkedTokenSource" since we have to cancel on a pair of cancelationTokens. An internally created one canceled with .CancelAfter(x) for the timeout, if configured, and the existing one coming from the caller for explicit .Cancel(). I'll work on it.

**Comment by i8beef** [comment]
HttpClient.Timeout (what should have been set using that _timeout that wasn't used) actually does that under the hood. Per request timeouts was always an issue wit hHttpClient, but I don't think we need that here, and it would be much simpler.

**Comment by Perecli** [comment]
So I changed line 18 to the following, to check how HttpClient handles the timeout.
```
private static HttpClient _httpClient = new HttpClient() { Timeout = TimeSpan.FromMilliseconds(1) };
```
It throws "A task was canceled." exception which means that internally they are doing the same thing I was going to do using a CancellationTokenSource(timeout) but the exception is not very descriptive about what actually happened.

What I wanted to do is similar but since we would control the code we can provide a better Timeout Exception with a better message.

"_httpClient" takes a "CancellationToken". So using a new CancellationTokenSource(timeout) as the parent of the consumer's CancellationToken (the child) and binding them with "CancellationTokenSource.CreateLinkedTokenSource" we could halt code execution on both the given CancellationToken and a timeout TokenSource but we can check which one threw the error and raise a better exception indicating what was the cause of the halt.

Do you have a simpler solution?

**Comment by Perecli** [comment]
Implemented new timeout feature. Works well in my tests.

**Comment by i8beef** [comment]
Ok, I'm like 50/50 here (other option being just using the native HttpClient.Timeout), but I think ultimately I like your solution in general as it allows a caller to discern a timeout from some other cancellation exception easier... You're just gonna make me go update all my other thin HttpClient libraries to handle timeout the same way :-P

So some notes:

1. Timeout should be handed through on the constructor with a default value.
2. Default timeout is low. I think we should keep the default at 100 seconds (to match default HttpClient). Remember, a DNS resolution can take up to 15 seconds alone before it times out, so I personally wouldn't go below 30 unless I was really sure about my environment... and as this is a DEFAULT we should err on the side of caution.
3. Timeout.InfiniteTimeSpan instead of arbitrary AddDays for default HttpClient timeout.
4. HttpClientSendAsyncWithTimeout is a little long, and though I'm not adverse to verbosity, I think I'd rather see just "SendAsync" or "SendAsyncWithTimeout". Also, obviously this would need to change to a non-static given point 1.

**Comment by Perecli** [comment]
1. The constructor is for instance members. The timeout is static. I wish it wasn't but it had to be because its called by a chain of other static methods. I probably would have had no static methods in the entire Client class but that was the original design I had to work with.

2. OK, Changed.

3. Hmm... tried TimeSpan.MaxValue and it threw an exception. Seems to work with Timeout.InfiniteTimeSpan so I changed it.

4. Changed the name. 
I could not make "SendAsyncWithTimeout" non-static because it is called by two static methods and didn't want to totally refactor the code. This is where I believe we should pick one or the other for all members. I favor non static.

**Comment by i8beef** [comment]
Yeah, this likely requires a larger refactor. Let me work on it a bit, I have some ideas that would just move to using the get and set Funcs completely and dump those statics. That would move the authToken management completely outside of the library, but let it still handle validation / refresh tokens. I hav ea proof of concept that feels alright, and I'll incorporate your timeout functionality into it.

**Comment by i8beef** [comment]
Check out that branch. It eliminates the need for static initialization pieces in favor of allowing a caller to completely manage the the auth token state externally. It's a little more complicated for callers, but this would handle BOTH cases in a single approach, and lets us push the timeout in the constructor as an optional parameter in a way that feels more natural.

**Comment by Perecli** [comment]
I like what you did. I would have refactored very similarly. Here are some small things I noticed:

1. I think its more flexible for the consumer to enter the timeout as a TimeSpan rather than seconds, that way they can benefit from the TimeSpan builders to create a timeout rather than having to calculate it in seconds. That's why Microsoft also accepts a TimeSpan for their timeout implementation.

2. You took the .Invoke out of _getStoredAuthTokenFunc.Invoke() but not out of _setStoredAuthTokenFunc. I usually code in VB.NET where the Invoke on a delegate is required so I didn't know you can omit it in C# but if you can, then be consistent throughout. 

3. Missing some .ConfigureAwait(false) on some of the _getStoredAuthTokenFunc and _setStoredAuthTokenFunc calls if you care about that.

4. In the GetAsync and PostAsync when you get the result from GetCurrentAuthTokenAsync, I think the variable receiving the result should be called "storedAuthToken" not "authToken" because "authToken" is the other class that is deserialized from Ecobee and this one is not that. It makes it confusing when you read the code if you are not consistent with other parts of the code where "authToken" does indeed reference the correct class. It confused me reading it.

5. To make things more bullet proof maybe we should throw exceptions in the constructor if the get/set delegate functions are null since in the code now they are invoked without the null check and because they are now required.

6. In "Program.cs" if a "token.txt" file exists the _currentAuthToken shown through Console.WriteLine() is blank since the ReadTokenFileAsync is not called until the first client.GetAsync() which comes later. Maybe those three Console.WriteLine() should be shown only if the file doesn't exist.

7. Look at the Remarks in this link: https://msdn.microsoft.com/en-us/library/system.threading.tasks.task.getawaiter(v=vs.110).aspx

**Comment by i8beef** [comment]
1. I did too, but as a default parameter Timespan is a struct and not compile time constant so it can't be used that way. I'll add a second constructor instead to handle that.
2. Fixed.
3. Fixed
4. Valid. Changed.
5. Good idea. Added.
6. Good catch.
7. Despite that remark, GetAwaiter().GetResult() has been in fairly common usage. It's main benefit is that it unwraps the AggregateException that Result will throw. However honestly, I should stop being lazy and just implement a MainAsync wrapper in that test app. I'll do that.

**Comment by Perecli** [comment]
Regarding number 7, yes I read that too in the forums about AggregateException 

I did the following in my console app test harness since the project would not allow me to make the Main async.
```
    Sub Main()
        Go().Wait()
    End Sub

    Private Async Function Go() As Task
        ...
        'all your code demoed with proper Awaits here
        await Whatever()
        ...
    End Function
```
Looks like you did a similar thing.

**Comment by Perecli** [comment]
```
<param name="timeout">Timeout for all HTTP calls in seconds, default 100.</param>
```
Is no longer in seconds so text should change.

**Comment by Perecli** [comment]
Everything looks good to me now.

**Comment by i8beef** [comment]
Ill make that final doc changes and get this merged then as 2.2.0 then.
